### ISSUE TITLE
[Rework] Add checks on workspace params

### ISSUE THREAD

**Issue by junliume** [ROCmSoftwarePlatform/MIOpen#2539]
@amberhassaan I have to revert #2498 for now but the work is still valuable even critical so we would like to update that feature branch further before we can merge another PR on it
CC: @JehandadKhan @atamazov 

**[Observations}**:
CI: http://micimaster.amd.com/blue/organizations/jenkins/MLLibs%2FMIOpen/detail/develop/1265/pipeline
Also observed on `gfx1100`:
[LastTest.log](https://github.com/ROCmSoftwarePlatform/MIOpen/files/13399562/LastTest.log)

It appears that the value of `workSpaceSize` is **not updated** from the latest **evaluated solver**:
- In this case, after "`[FindSolutionImpl] ConvDirectNaiveConvWrw`" with "`[ValidateWorkspace] workSpace = 0x7f6ea4607000, workSpaceSize = 8192`", eventually the chosen one "`[FindConvFwdAlgorithm] FW Chosen Algorithm: ConvDirectNaiveConvFwd`" which should have "`[ValidateWorkspace] workSpace = 0, workSpaceSize = 0`"
- However, `workSpace` seems to be updated while `workSpaceSize` is not updated from the last value, so it stayed at `8192` in "`[ValidateWorkspace] workSpace = 0, workSpaceSize = 8192`"
- This will cause the workSpace check to fail

**Comment by amberhassaan** [comment]
@junliume : I'll take a look. It seems the check did its job :). I have a fix in progress here and I'll make sure it handles the case above too: https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2524

**Comment by atamazov** [comment]
[Notice] @junliume 
> the work is still valuable even critical

This is definitely valuable! but I am not sure if this is critical or not. Maybe there are (or were) some P0 tickets related to incorrect workspace ptr/size passed by the client to the library?

> I have to revert https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2498 for now

That's exactly the case I meant in https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2498#issuecomment-1802910907.
